packages/kbn-pm/src/run.ts

```diff
@@ -24,6 +24,31 @@ export async function runCommand(command: Command, config: CommandConfig) {
     );
 
     const projects = await getProjects(config.rootPath, projectPaths);
+    const exclude = getExcludeIncludeFilter(config.options.exclude);
+    const include = getExcludeIncludeFilter(config.options.include);
```

Comment:
What do you think about moving this logic into `getProjects`? E.g. something like this:

```diff
diff --git a/packages/kbn-pm/src/run.ts b/packages/kbn-pm/src/run.ts
index 43cd6e9248..76729c25f7 100644
--- a/packages/kbn-pm/src/run.ts
+++ b/packages/kbn-pm/src/run.ts
@@ -23,22 +23,10 @@ export async function runCommand(command: Command, config: CommandConfig) {
       config.options as ProjectPathOptions
     );
 
-    const projects = await getProjects(config.rootPath, projectPaths);
-    const exclude = getExcludeIncludeFilter(config.options.exclude);
-    const include = getExcludeIncludeFilter(config.options.include);
-
-    // Filter out projects that shouldn't be included if any filters are specified.
-    if (exclude.length > 0 || include.length > 0) {
-      for (const [projectName] of projects) {
-        const excludeProject =
-          exclude.includes(projectName) ||
-          (include.length > 0 && !include.includes(projectName));
-
-        if (excludeProject) {
-          projects.delete(projectName);
-        }
-      }
-    }
+    const projects = await getProjects(config.rootPath, projectPaths, {
+      include: toArray(config.options.include),
+      exclude: toArray(config.options.exclude),
+    });
 
     if (projects.size === 0) {
       console.log(
@@ -82,14 +70,14 @@ export async function runCommand(command: Command, config: CommandConfig) {
   }
 }
 
-function getExcludeIncludeFilter(excludeIncludeRawValue?: string | string[]) {
-  if (excludeIncludeRawValue == null) {
+function toArray(value?: string | string[]) {
+  if (value == null) {
     return [];
   }
 
-  if (typeof excludeIncludeRawValue === 'string') {
-    return [excludeIncludeRawValue];
+  if (typeof value === 'string') {
+    return [value];
   }
 
-  return excludeIncludeRawValue;
+  return value;
 }
diff --git a/packages/kbn-pm/src/utils/projects.ts b/packages/kbn-pm/src/utils/projects.ts
index b465b56531..31b4687ce1 100644
--- a/packages/kbn-pm/src/utils/projects.ts
+++ b/packages/kbn-pm/src/utils/projects.ts
@@ -9,10 +9,12 @@ const glob = promisify(globSync);
 
 export type ProjectMap = Map<string, Project>;
 export type ProjectGraph = Map<string, Project[]>;
+export type ProjectsOptions = { include?: string[]; exclude?: string[] };
 
 export async function getProjects(
   rootPath: string,
-  projectsPathsPatterns: string[]
+  projectsPathsPatterns: string[],
+  { include = [], exclude = [] }: ProjectsOptions = {}
 ) {
   const projects: ProjectMap = new Map();
 
@@ -24,6 +26,12 @@ export async function getProjects(
       const projectDir = path.dirname(projectConfigPath);
       const project = await Project.fromPath(projectDir);
 
+      const excludeProject =
+        exclude.includes(project.name) ||
+        (include.length > 0 && !include.includes(project.name));
+
+      if (excludeProject) continue;
+
       if (projects.has(project.name)) {
         throw new CliError(
           `There are multiple projects with the same name [${project.name}]`,
```

(I also renamed to `toArray` just because I felt the intention was clearer)

Reply:
With this change we can also move more of the tests over to the `getProjects` tests.

Reply:
Oh, an improved `toArray` would of course be:

```ts
function toArray<T>(value?: T | T[]) {
  if (value == null) {
    return [];
  }

  if (Array.isArray(value)) {
    return value;
  }

  return [value];
}```

Reply:
> What do you think about moving this logic into getProjects?

Yeah, it looks much better! Initially I was thinking about use cases where we may need full/unfiltered list, but couldn't find any, so we should be good :)

Reply:
> With this change we can also move more of the tests over to the getProjects tests.

Yeah, but still have to keep a bunch of tests in `run.tests` to test string/string[] behavior for exclude/include filters :/